File: backend/app/schemas/message.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class MessageRead(BaseModel):
    id: uuid.UUID
    match_id: uuid.UUID
    sender_id: uuid.UUID
    content: str
    message_type: str
    is_read: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class ConversationRead(BaseModel):
    """Une conversation = un match. Format attendu par l'app mobile."""

    room_id: str
    other_user_name: str
    other_user_avatar: str | None = None
    last_message: str | None = None
    last_at: str | None = None
    unread: int = 0


class WsIncoming(BaseModel):
    content: str
    message_type: str = "text"


class WsOutgoing(BaseModel):
    id: str
    match_id: str
    sender_id: str
    content: str
    message_type: str
    created_at: str
    is_read: bool = False

File: mobile/lib/features/messaging/models/message.dart
class Conversation {
  final String roomId;
  final String otherUserName;
  final String? otherUserAvatar;
  final String? lastMessage;
  final DateTime? lastAt;
  final int unread;

  const Conversation({
    required this.roomId,
    required this.otherUserName,
    this.otherUserAvatar,
    this.lastMessage,
    this.lastAt,
    this.unread = 0,
  });

  factory Conversation.fromJson(Map<String, dynamic> j) => Conversation(
        roomId: j['room_id'] as String,
        otherUserName: j['other_user_name'] as String,
        otherUserAvatar: j['other_user_avatar'] as String?,
        lastMessage: j['last_message'] as String?,
        lastAt: j['last_at'] != null
            ? DateTime.parse(j['last_at'] as String)
            : null,
        unread: (j['unread'] as int?) ?? 0,
      );
}

class ChatMessage {
  final String id;
  final String roomId;
  final String senderId;
  final String content;
  final DateTime createdAt;
  final bool isMe;
  final bool isRead;

  const ChatMessage({
    required this.id,
    required this.roomId,
    required this.senderId,
    required this.content,
    required this.createdAt,
    required this.isMe,
    this.isRead = false,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> j, String myId) =>
      ChatMessage(
        id: j['id'] as String,
        // Le backend renvoie `match_id` (un match = une room de chat).
        roomId: (j['room_id'] ?? j['match_id']) as String,
        senderId: j['sender_id'] as String,
        content: j['content'] as String,
        createdAt: DateTime.parse(j['created_at'] as String),
        isMe: j['sender_id'] == myId,
        isRead: j['is_read'] as bool? ?? false,
      );
}
